cuqi/utilities/_utilities.py

```diff
@@ -173,18 +163,45 @@ def approx_derivative(func, wrt, direction=None, epsilon=np.sqrt(np.finfo(np.flo
     # corresponding geometry par2fun map is not identity
     # (e.g. funvalues=paramters**2), because wrt entries 
     # are interpreted as function value.
-    if isinstance(wrt, CUQIarray) or isinstance(direction, CUQIarray):
-        raise NotImplementedError("approx_derivative is not implemented"+
-                                  "for inputs of type CUQIarray")
+
+    if isinstance(wrt, CUQIarray):
+        wrt = wrt.to_numpy()
+    if direction is not None and isinstance(direction, CUQIarray):
+        direction = direction.to_numpy()
+    
     wrt = np.asfarray(wrt)
     f0 = func(wrt)
-    Matr = np.zeros([len(wrt), len(f0)])
+    Matr = np.zeros([infer_len(wrt), infer_len(f0)])
     dx = np.zeros(len(wrt))
     for i in range(len(wrt)):
         dx[i] = epsilon
         Matr[i] = (func(wrt+dx) - f0)/epsilon
         dx[i] = 0.0
     if direction is None:
-        return Matr.T
+        if infer_len(f0) == 1:
+            return Matr.reshape(infer_len(wrt))
+        else:
+            return Matr.T
     else:
-        return Matr@direction
\ No newline at end of file
+        return Matr@direction
+
+def approx_gradient(func, x, epsilon= 0.000001):
+    """Approximates the gradient of callable scalar function `func` evaluated at point `x`. The approximation is done using finite differences with
+    step size `epsilon`."""
+
+    #TODO: use approx_derivative after making it data structure agnostic
+    #return approx_derivative(func, x, None, epsilon=epsilon)
+ 
+    if isinstance(x, Number):
+        return (func(x+epsilon) - func(x))/epsilon
+
+    FD_gradient = copy.deepcopy(x)*0.0
+    eps_vec = copy.deepcopy(x)*0.0
+
+    for i, x_i in enumerate(x):
```

Comment:
Not sure if all arrays are enumerateble like that. Perhaps a simply len(x) and loop over that is better?

Reply:
Good point, I ended up using infer_len because __len__ is also not necessarily implanted I think, based on array API specification: https://data-apis.org/array-api/latest/API_specification/index.html

Reply:
Great thanks!